### ISSUE TITLE
Printing problems on Linux

### ISSUE THREAD

**Issue by cocoabsb** [wxWidgets/wxWidgets#24229]
I am using wxWidgets-3_2_4, Mac OS X 12.7, Windows 10 and Linux Mint 21.2 and gcc for Mac OS X and LInux and Visual C++ for Windows 10.

I am noticing that on Mac and Windows the printing system on wxWidgets works flawlessly but on Linux not all characters on a page are printed when I send the report to the printer. When I print the report to a pdf file and then send the file to the printer, all characters are printed flawlessly but when I send the report directly to the printer, many characters are not printed.

I tried the printing sample for wxWidgets-3_2_4 and to my surprise the Page number is NOT being printed for Linux either and many characters are printed bigger than on Mac or Windows. That happens only if I send the printing sample directly to the printer. On Mac and Windows, even if I send the printing sample directly to the same printer the Page number is printed flawlessly and the other texts are printed flawlessly.

I am attaching scanned images of the printing sample for Linux and Mac so you will see that for Linux the page number is not being printed and the other characters are being printed bigger, although the source code is exactly the same for the three operating systems.

All the graphics are being printed correctly, just like they are being printed on Mac and Windows, but the texts are printed bigger and sometimes apparently at wrong coordinates.

It appears that on Linux the coordinates for drawing text are wrong. The PAGE number, for instance, should be printed at 0,0 coordinates, but on Linux it does NOT get printed correctly, only some dashes are printed, as if the 0,0 coordinates are NOT 0,0, but some negative numbers, above the top left corner of the page.

It is possible to simulate the issue just by downloading wxWidgets-3.2.4, compiling it and then compiling the printing sample, running it and clicking on the File->Print menu option. That's exactly what I did.

#### Bug description:

The printing features of wxWidgets (at least release 3.1.2 and 3.2.4) are not being printed correctly on Linux only: the characters are bigger than on Mac and/or Windows and at wrong coordinates. Only texts are being printed incorrectly and only on Linux. The graphics are being printed correctly for the three operating systems.

#### Expected vs observed behaviour:

I used the printing sample (which is part of wxWidgets) and the expected behaviour is that for Mac, Windows and Linux the printing sample should be printed the same way and with no issues. The observed behaviour is that for Mac and Windows the printing sample is being printed the same way and correctly, but for Linux the characters (text characters) are being printed incorrectly, either bigger than on Mac or Windows or at wrong coordinates having even a few characters which are NOT being printed at all, as for example the PAGE number on the wxWidgets' printing sample.

#### To Reproduce:

1. Download wxWidgets-3.2.4 for Linux
2. Compile it
3. Go to the printing sample folder/directory
4. Compile it
5. Run it on Linux (./printing)
6. Click on File-\>Print
7. Send it to the Printer (NOT to a file)
8. See error

### Platform and version information

- wxWidgets version: [3.2.4]
- wxWidgets port: [wxGTK]
- OS: [Linux Mint 21.2]
  
  + GTK version: [3.24.33] 
  + Which GDK:  [X11]
  + Desktop environment: [Gnome]
  + Current theme: [Cinnamon]
[2024-01-18-Printing_Sample_Linux-Page1.pdf](https://github.com/wxWidgets/wxWidgets/files/13992519/2024-01-18-Printing_Sample_Linux-Page1.pdf)
[2024-01-18-Printing_Sample_MacOSX-Page1.pdf](https://github.com/wxWidgets/wxWidgets/files/13992521/2024-01-18-Printing_Sample_MacOSX-Page1.pdf)
<!--

**Comment by vadz** [comment]
If PDF output under Linux is correct, it looks more like a printer driver problem than anything  else, doesn't it?

**Comment by cocoabsb** [comment]
Then why if I print it to a file and send the file to the printer, it is printed flawlessly?

**Comment by dsa-t** [comment]
"Text by wxGraphicsContext" depends on system DPI scaling. Don't know if this is intentional or not.

Checking on 3 printers, the main rectangle is centered on all of them, but the page number is out of printing bounds on 2/3 printers. On the other printer, the bottom part of the page text is slightly visible.

I don't know what size the other text should be. Maybe the font is different.

Printing to PDF, then on printer with "Actual size" scaling from msedge produce exact same results as printing directly from wx sample.

Linux Mint 21.1.

**Comment by vadz** [comment]
> "Text by wxGraphicsContext" depends on system DPI scaling. Don't know if this is intentional or not.

No, this is definitely a bug :-(
 
> Checking on 3 printers, the main rectangle is centered on all of them, but the page number is out of printing bounds on 2/3 printers. On the other printer, the bottom part of the page text is slightly visible.

This looks like a bug too...

**Comment by cocoabsb** [comment]
I decided to use the printing sample from wxWidgets-3.1.2 on Linux, printed to a file and sent the file to the printer and the output was exactly the same as on Mac. If though I send it directly to the printer, the issue happens just as described above and exactly the same way as in wxWidgets-3.2.4. 

I noticed a difference from the printing sample output on linux from wxWidgets-3.1.2 and wxWidgets-3.2.4 if I print it to a file and then send the file to the printer: In wxWidgets-3.1.2, the output is the same as on Mac and Windows whereas in wxWidgets-3.2.4, there are a few differences.

I am attaching pdf files from the printing sample output, both for Linux, when I print it to a file and send the file to a printer so you can see the differences between the output from wxWidgets-3.1.2 and wxWidgets-3.2.4.

[2024-01-19-Printing_Sample_Linux_from_pdf_file-wxWidgets-3_1_2.pdf](https://github.com/wxWidgets/wxWidgets/files/13994940/2024-01-19-Printing_Sample_Linux_from_pdf_file-wxWidgets-3_1_2.pdf)
[2024-01-19-Printing_Sample_Linux_from_pdf_file-wxWidgets-3_2_4.pdf](https://github.com/wxWidgets/wxWidgets/files/13994941/2024-01-19-Printing_Sample_Linux_from_pdf_file-wxWidgets-3_2_4.pdf)

**Comment by dsa-t** [comment]
>I decided to use the printing sample from wxWidgets-3.1.2 on Linux ... If though I send it directly to the printer, the issue happens just as described above

I don't think the text size will be as big as in 3.2.4 in this case. Can you attach screenshot of a preview from the system printing dialog?

**Comment by cocoabsb** [comment]
There you go
![Preview-wxWidgets-3 1 2](https://github.com/wxWidgets/wxWidgets/assets/57816895/48fb61b6-fc49-4296-9b4c-dca71db14545)
wxWidgets-3.1.2
![Preview-wxWidgets-3 2 4](https://github.com/wxWidgets/wxWidgets/assets/57816895/da8d558d-31e3-49c7-86b4-647b621764c8)
wxWidgets-3.2.4

**Comment by MaartenBent** [comment]
To me it doesn't look like a system DPI scaling problem, but more like a 72ppi vs 96ppi problem, with text 33% too big.
Because I also see the difference in Ubuntu at 100%/96dpi.

For the print-preview this could be fixed in the `wxGraphicsContext` cairo renderer by correcting for this factor in:
https://github.com/wxWidgets/wxWidgets/blob/f37401dde3f359860a08236fe8d6414b7c93ee88/src/generic/graphicc.cpp#L2547
with this ugly fix:
```c++
m_fontScalingFactor = screen ? float((gdk_screen_get_resolution(screen) / 96.0) * 72.0/96.0) : 1.0f;
```

But this would mean currently all text in cairo graphics contexts is too big?


For printing itself (at least to PDF), the fix is nicer. It seems we can simply remove the `GetFontPointSizeAdjustment(72.0)` calls in `src/gtk/print.cpp` (and the related `pango_font_description_get_size`/`pango_font_description_set_size` calls).

I don't know if this is related to the page number and coordinate problems, probably not.
With issues like this it is also useful to try to reproduce it at 100% display scaling, to figure out if the display scale factor causes any of these problems.

> "Text by wxGraphicsContext" depends on system DPI scaling. Don't know if this is intentional or not.

Not sure what you mean here? If you mean text at 14pt appears twice as big at 200% scaling vs 100% scaling, then this is intentional.

**Comment by dsa-t** [comment]
>Not sure what you mean here? If you mean text at 14pt appears twice as big at 200% scaling vs 100% scaling, then this is intentional.

96 DPI in Mint Appearance - Fonts - Custom DPI setting
![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/18d92fe0-80f6-4292-b4d8-69cdd49db441)

130 DPI
![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/1c04ce17-94c6-4858-ac7f-213b19def474)

**Comment by MaartenBent** [comment]
Yeah, that's not what it is supposed to look like. Only the print preview, or also the main screen?
The weird thing is that both wxDC and wxGraphicsContext use a cairo graphics context internally in gtk3. Though the text in the wxDC is fine.

**Comment by dsa-t** [comment]
96:
![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/f71b4af4-1a3a-4773-b1c3-74a4234a2e0c)

130:
![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/9899c583-e7ef-4817-9e30-36a37b2b9d9a)

**Comment by MaartenBent** [comment]
Looking at the history of that `m_fontScalingFactor`, this might be by design. It is supposed to ensure that generic widgets have their text scaled appropriately to be consistent with native widgets: fbc75ec8e012d2eff3ab755a77dd3ba6c2964a26
But the consequence is that text on every cairo graphics context is scaled. Then the question is, why is not all text in the 130 DPI print preview bigger...
